### ISSUE TITLE
Nested volume mounts are unexpectedly locked in 3.4.0

### ISSUE THREAD

**Issue by pvande** [docker/for-mac#5748]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Expected behavior

In 3.3.3 and prior versions, given a simple project:

```
$ tree
.
├── cached
├── shared
├── unsynced
└── docker-compose.yaml

3 directories, 1 file
```

… with a simple configuration file:

``` yaml

version: "2.2"
services:
  app:
    image: alpine
    volumes:
      - .:/app/:delegated
      - ./cached:/app/cached/:cached
      - shared:/app/shared/:cached
      - /app/unsynced/
    working_dir: /app

volumes:
  shared: {}
```

Each of the three directories were writable on both the host machine and the container.

### Actual behavior

In 3.4.0, both the `cached` and `shared` volume mounts are marked as "Locked" by the OS (visible in Finder, and by `ls -lO`).

```
# Project not running
$ ls -lO 
total 8
drwxr-xr-x  2 user  staff  -  64 Jun 11 22:56 cached
-rw-r--r--  1 user  staff  - 224 Jun 11 23:01 docker-compose.yaml
drwxr-xr-x  2 user  staff  -  64 Jun 11 22:56 shared
drwxr-xr-x  2 user  staff  -  64 Jun 11 22:56 unsynced

# Project is running
$ ls -lO
total 8
drwxr-xr-x  2 user  staff  uchg  64 Jun 11 22:56 cached
-rw-r--r--  1 user  staff  -    224 Jun 11 23:01 docker-compose.yaml
drwxr-xr-x  2 user  staff  uchg  64 Jun 11 22:56 shared
drwxr-xr-x  2 user  staff  -     64 Jun 11 22:56 unsynced
```

This makes the directories unwritable from the host, and makes the `cached` directory unwritable from the container.

### Information
  - macOS Version: Reproduced in Catalina (10.15.7) and Big Sur/
  - Intel chip or Apple chip: Intel
  - Docker Desktop Version: 3.4.0

This behavior has been reproduced on two separate machines, and has been validated as working when downgrading to v3.3.3.

### Steps to reproduce the behavior

  1. Create a simple project with nested volume mounts
  2. Start the project (e.g. `docker-compose run --rm app sh`)
  3. Observe that file writes to the nested volume from both the host and the container fail, and that the nested volume is locked against user writes.

**Comment by wkirby** [comment]
I'm not 100% sure this is the issue I was experiencing after upgrading — I had to roll back immediately to retain any productivity — but I was very reliably unable to propogate changes from the host to the container, in either the `cached` or `default` configuration.

Something has definitely regressed in 3.4 with regard to volume mounts.

**Comment by wmuizelaar** [comment]
I had a situation where a nested volume mount didn't work as well:

```
docker run <<snip other docker run commands>>
    -v /Users/wmuizelaar/app/data:/app/data:ro
    -v /Users/wmuizelaar/app/data/more_specific:/app/data/more_specific:delegated
 ```
 
In the previous versions the more specific volume was writeable (and that was the intention as well), but after upgrading to 3.4.0 this didn't work anymore, and everythong was readonly.

**Comment by stephen-turner** [comment]
Thanks for the report. We think we know what caused this: it was a side-effect of another bug fix. Assigning to @djs55 to see what we can do.

**Comment by djs55** [comment]
@pvande @wkirby @wmuizelaar I've got a candidate fix for this issue if you'd like to try it: https://desktop-stage.docker.com/mac/stable/amd64/65680/Docker.dmg (replace `amd64` with `arm64` for Apple Silicon) In this build the directories are not locked and the contents remain writable and deletable. If you get a chance to try it, let me know how it goes. If it still doesn't work, let me know the details and I'll investigate. Thanks in advance!

**Comment by rodcarroll** [comment]
I ran into this as well, and was able to work around by disabling "Use gRPC FUSE for file sharing"

**Comment by wmuizelaar** [comment]
@djs55 I can confirm that this version fixes the issue.

**Comment by wmuizelaar** [comment]
Side-effect with this version though is that the UI states 'No containers running' while `docker ps` strongly disagrees (I have a docker-compose setup running currently)

**Comment by djs55** [comment]
@wmuizelaar regarding the docker ps / dashboard issue, could you upload some diagnostics? I'm curious about that one

**Comment by wmuizelaar** [comment]
@djs55 A976B560-4FFD-4903-8D4A-777CF3BA8D09/20210616090901 is the diagnostics-ID

**Comment by wmuizelaar** [comment]
Not sure if related, but I had to manually correct the docker-compose symlink to get it working again, see also https://github.com/docker/for-mac/issues/5593

**Comment by djs55** [comment]
@wmuizelaar Thanks for the diagnostics! I think you're right, the symlink code isn't being run on upgrade from 3.4.0 in the latest development build. I'll get this fixed before we ship a 3.5.0 😄

**Comment by stevebutler2210** [comment]
We are experiencing behaviour very similar to this in one of our projects. At a stripped back level our setup looks like something like this

```
version: "3.4"
services:
  myapp:
    build:
      context: ./myapp
      dockerfile: Dockerfile.dev
    image: "myapp:master"
    volumes:
      - app-mount:/app:cached
      - other-app-mount:/app/tmp/other-app:cached
      - private-mount:/app/private
      - shared-bundle-cache:/bundle
      - ${PWD}/docker-entrypoint.sh:/docker-entrypoint.sh
      - /run/host-services/ssh-auth.sock:/run/host-services/ssh-auth.sock
    depends_on:
      - db
      - selenium

x-volumes: &default-volume
  driver: local

x-driver-opts: &default-driver-opts
  type: none
  o: bind

volumes:
  app-mount:
    <<: *default-volume
      driver_opts:
        <<: *default-driver-opts
        device: "${PWD}/myapp"
  other-app-mount:
    <<: *default-volume
      driver_opts:
        <<: *default-driver-opts
        device: "${PWD}/other-app"
  private-mount:
    <<: *default-volume
    driver_opts:
      <<: *default-driver-opts
      device: "${PWD}/data"
```

Prior to upgrading to 3.4.0 this worked with no issues. However, the files are now being mounted as read-only (confirmable via `ls -l`

```
/app# ls -l spec/my/test.rb
-rw-r--r-- 1 root root 4606 Apr 12 08:56 spec/my/test.rb
```

We _are_ still able to write to the files on the host machine, but these changes aren't reflected in the container without re-building. Somewhat strangely, even deleting and re-creating the volume with `docker volume rm` doesn't seem to help.

Using `docker container inspect` to compare the before and after does highlight the following unexpected difference in the volume mount config:
```
# before

$ docker container inspect myapp-container
# ...
{
  "Type": "volume",
  "Name": "myapp-mount",
  "Source": "/var/lib/docker/volumes/myapp-mount/_data",
  "Destination": "/app",
  "Driver": "local",
  "Mode": "cached",
  "RW": true,
  "Propagation": ""
},
# ...

# after
$ docker container inspect myapp-container
# ...
{
  "Type": "volume",
  "Name": "myapp-mount",
  "Source": "/var/lib/docker/volumes/myapp-mount/_data",
  "Destination": "/app",
  "Driver": "local",
  "Mode": "z",
  "RW": true,
  "Propagation": ""
},
# ...
```

The `docker-compose.yml` is unchanged from the working configuration, so I guess the mount mode changing is likely a symptom of this bug?

**Comment by StefanScherer** [comment]
We have just released Docker Desktop 3.5.0 which contains a fix for this issue. I'll close this ticket now, but feel free to comment and give us feedback. If there is still a problem, please open a new ticket.
